be/dtm0_log.h

```diff
@@ -0,0 +1,522 @@
+	struct m0_mutex          dl_lock;
+	struct m0_be_seg	*dl_seg;
+	/** DTM0 clock source */
+	struct m0_dtm0_clk_src	*dl_cs;
+	union {
+		/** Persistent list, used if dl_is_persistent */
+		struct m0_be_list	*dl_list;
+		/** Volatile list, used if !dl_is_persistent */
+		struct m0_tl		*dl_tlist;
+	} u;
+};
+
+/**
+ * @b Typical call flow for calling DTM0 log interface routines:
+ *
+ * ** Initialisation/Finalisation for volatile fields of the log
+ * ** Create a persistent log or just allocate a volatile (in-memory) log
+ *   m0_be_dtm0_log_create / m0_be_dtm0_log_alloc
+ * - m0_be_dtm0_log_init()
+ * ** Preparation phase
+ * - m0_be_dtm0_log_credit()
+ * ** Normal operation phase
+ * - m0_be_dtm0_log_update()
+ * - m0_be_dtm0_log_prune()
+ * - m0_be_dtm0_log_find()
+ */
+
+/**
+ * Allocate an m0_be_dtm0_log. This call will allocate a dtm0 volatile log.
+ * We need to call this routine once before we can perform any operations on .
+ *
+ * @pre out != NULL.
+ * @post *log is allocated and ready to be initialized.
+ *
+ * @param log Pointer to variable in which to return the allocate memory
+ *        address.
+ *
+ * @return 0 on success. Anything else is a failure.
+ */
+M0_INTERNAL int m0_be_dtm0_log_alloc(struct m0_be_dtm0_log **log);
+
+/**
+ * Initialize a dtm0 log. We need to call this routine once before we can
+ * perform any operations on it.
+ *
+ * @pre log != NULL;
+ * @pre cs != NULL;
+ * @post *log is initialized and ready to be used
+ *
+ * @param log Pointer to a m0_be_dtm0_log that has to be initialized.
+ * @param cs Pointer to a clock source. This will be used for comparisons
+ *        involving log record timestamps.
+ * @param isvstore A flag to indicate whether the dtm0 log that we are trying
+ *        to init is a volatile of persistent store.
+ *
+ * @return 0 on success. Anything else is a failure.
+ */
+M0_INTERNAL int m0_be_dtm0_log_init(struct m0_be_dtm0_log  *log,
+                                    struct m0_dtm0_clk_src *cs,
+                                    bool                    is_plog);
+
+/**
+ * Finalize a dtm0 log. We need to call this routine once after we have
+ * finished performing all operations on it.
+ *
+ * @pre m0_be_dtm0_log__invariant needs to be satisfied.
+ * @post All fields of *log are finalized. For volatile log the memory
+ *       allocated by m0_be_dtm0_log_alloc is freed.
+ *
+ * @param log Pointer to a valid log.
+ *
+ * @return None
+ */
+M0_INTERNAL void m0_be_dtm0_log_fini(struct m0_be_dtm0_log **log);
+
+/**
+ * Free the memory allocated by m0_be_dtm0_log_alloc
+ *
+ * @pre m0_be_dtm0_log->dl_is_persistent needs to be false.
+ * @post *log is set to NULL.
+ *
+ * @param log Pointer to a log structure that has been previously allocated.
+ *
+ * @return None
+ */
+M0_INTERNAL void m0_be_dtm0_log_free(struct m0_be_dtm0_log **log);
+
+/**
+ * For performing an operation on a persistent log, we need to take the
+ * appropriate number of be transaction credit.
+ *
+ * @pre None
```

Comment:
We can remove the "@pre" condition if it is None.
But actually it is not None. At least these arguments must be valid.

Reply:
OK, Done.